Applications/Spire/Source/Ui/TimeInForceFilterPanel.cpp

```diff
@@ -0,0 +1,49 @@
+#include "Spire/Ui/TimeInForceFilterPanel.hpp"
+#include "Spire/Ui/ArrayListModel.hpp"
+#include "Spire/Ui/ArrayTableModel.hpp"
+
+using namespace Nexus;
+using namespace Spire;
+
+namespace {
+  std::unordered_map<TimeInForce::Type, int>& populate() {
+    static auto values = [] {
+      auto values = std::unordered_map<TimeInForce::Type, int>();
+      auto i = 0;
+      values[TimeInForce::Type::DAY] = i++;
+      values[TimeInForce::Type::GTC] = i++;
+      values[TimeInForce::Type::OPG] = i++;
+      values[TimeInForce::Type::IOC] = i++;
+      values[TimeInForce::Type::FOK] = i++;
+      values[TimeInForce::Type::GTX] = i++;
+      values[TimeInForce::Type::GTD] = i++;
+      values[TimeInForce::Type::MOC] = i++;
+      return values;
+    }();
+    return values;
+  }
+}
+
+TimeInForceFilterPanel* Spire::make_time_in_force_filter_panel(
+    QWidget& parent) {
+  return make_time_in_force_filter_panel(
+    std::make_shared<TimeInForceListModel>(std::make_shared<ArrayListModel>()),
+      parent);
+}
+
+TimeInForceFilterPanel* Spire::make_time_in_force_filter_panel(
+    std::shared_ptr<TimeInForceListModel> selected_model, QWidget& parent) {
+  auto model = std::make_shared<ArrayTableModel>();
+  auto& values = populate();
+  for(auto& value : values) {
+    model->push({TimeInForce(value.first), false});
```

Comment:
Avoid doing the same work on every function invocation.

Reply:
The second column of the TableModel is variable for the different instances of the TimeInForceFilterPanel, so it is impossible to create a static pointer of the TableModel for different instances of TimeInForceFilterPanel. Besides, since the copy constructor of the TableModel is deleted, the static object of the TableModel can’t be copied for the different TimeInForceFilterPanels. Hence, I have to push data to the TableModel for a new TimeInForceFilterPanel instance.